Crash with .AstroidError when parsing a dict with a lambda as the key

### Bug description

pylint crashes over this code

test.py

```python
a = {
    (lambda x: print(x)): 'foo'
}
```

### Configuration

_No response_

### Command used

```shell
> pylint C:\test.py
```


### Pylint output

```shell
************* Module test
test.py:3:0: C0304: Final newline missing (missing-final-newline)
test.py:1:0: C0114: Missing module docstring (missing-module-docstring)
Exception on node <Dict.dict l.1 at 0x1e448164df0> in file 'C:\test.py'
Traceback (most recent call last):
  File "C:\Python310\site-packages\pylint\utils\ast_walker.py", line 90, in walk
    callback(astroid)
  File "C:\Python310\site-packages\pylint\checkers\typecheck.py", line 2036, in visit_dict
    if not is_hashable(k):
  File "C:\Python310\site-packages\pylint\checkers\utils.py", line 1945, in is_hashable
    hash_fn = next(inferred.igetattr("__hash__"))
AttributeError: 'Lambda' object has no attribute 'igetattr'
test.py:1:0: F0002: test.py: Fatal error while checking 'test.py'. Please open an issue in our bug tracker so we address this [...]. (astroid-error)

-8<--------------------------------------------------------------------------

Traceback (most recent call last):
  File "C:\Python310\site-packages\pylint\lint\pylinter.py", line 782, in _lint_file
    check_astroid_module(module)
  File "C:\Python310\site-packages\pylint\lint\pylinter.py", line 1049, in check_astroid_module
    retval = self._check_astroid_module(
  File "C:\Python310\site-packages\pylint\lint\pylinter.py", line 1099, in _check_astroid_module
    walker.walk(node)
  File "C:\Python310\site-packages\pylint\utils\ast_walker.py", line 93, in walk
    self.walk(child)
  File "C:\Python310\site-packages\pylint\utils\ast_walker.py", line 93, in walk
    self.walk(child)
  File "C:\Python310\site-packages\pylint\utils\ast_walker.py", line 90, in walk
    callback(astroid)
  File "C:\Python310\site-packages\pylint\checkers\typecheck.py", line 2036, in visit_dict
    if not is_hashable(k):
  File "C:\Python310\site-packages\pylint\checkers\utils.py", line 1945, in is_hashable
    hash_fn = next(inferred.igetattr("__hash__"))
AttributeError: 'Lambda' object has no attribute 'igetattr'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Python310\site-packages\pylint\lint\pylinter.py", line 747, in _lint_files
    self._lint_file(fileitem, module, check_astroid_module)
  File "C:\Python310\site-packages\pylint\lint\pylinter.py", line 784, in _lint_file
    raise astroid.AstroidError from e
astroid.exceptions.AstroidError
```


### Expected behavior

No crash. :)

### Pylint version

```shell
pylint 2.15.2
astroid 2.12.9
Python 3.10.7 (tags/v3.10.7:6cc6b13, Sep  5 2022, 14:08:36) [MSC v.1933 64 bit (AMD64)]
```


### OS / Environment

Windows 10

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/7453.bugfix
+++ b/doc/whatsnew/fragments/7453.bugfix
@@ -1,0 +1,3 @@
+Fixed a crash in the ``unhashable-member`` checker when using a ``lambda`` as a dict key.
+
+Closes #7453

--- a/pylint/checkers/utils.py
+++ b/pylint/checkers/utils.py
@@ -19,7 +19,7 @@ from typing import TYPE_CHECKING, Callable, TypeVar
 
 import _string
 import astroid.objects
-from astroid import TooManyLevelsError, bases, nodes
+from astroid import TooManyLevelsError, nodes
 from astroid.context import InferenceContext
 from astroid.exceptions import AstroidError
 from astroid.nodes._base_nodes import ImportNode
@@ -37,15 +37,6 @@ COMP_NODE_TYPES = (
     nodes.SetComp,
     nodes.DictComp,
     nodes.GeneratorExp,
-)
-WITH_IGETATTR = (
-    nodes.Module,
-    nodes.ClassDef,
-    nodes.FunctionDef,
-    astroid.objects.Super,
-    nodes.Slice,
-    bases.UnboundMethod,
-    bases.BaseInstance,
 )
 EXCEPTIONS_MODULE = "builtins"
 ABC_MODULES = {"abc", "_py_abc"}
@@ -1959,7 +1950,7 @@ def is_hashable(node: nodes.NodeNG) -> bool:
         for inferred in node.infer():
             if inferred is astroid.Uninferable:
                 return True
-            if not isinstance(inferred, WITH_IGETATTR):
+            if not hasattr(inferred, "igetattr"):
                 return True
             hash_fn = next(inferred.igetattr("__hash__"))
             if hash_fn.parent is inferred:

--- a/pylint/checkers/utils.py
+++ b/pylint/checkers/utils.py
@@ -19,7 +19,7 @@ from typing import TYPE_CHECKING, Callable, TypeVar
 
 import _string
 import astroid.objects
-from astroid import TooManyLevelsError, nodes
+from astroid import TooManyLevelsError, bases, nodes
 from astroid.context import InferenceContext
 from astroid.exceptions import AstroidError
 from astroid.nodes._base_nodes import ImportNode
@@ -37,6 +37,15 @@ COMP_NODE_TYPES = (
     nodes.SetComp,
     nodes.DictComp,
     nodes.GeneratorExp,
+)
+WITH_IGETATTR = (
+    nodes.Module,
+    nodes.ClassDef,
+    nodes.FunctionDef,
+    astroid.objects.Super,
+    nodes.Slice,
+    bases.UnboundMethod,
+    bases.BaseInstance,
 )
 EXCEPTIONS_MODULE = "builtins"
 ABC_MODULES = {"abc", "_py_abc"}
@@ -1949,6 +1958,8 @@ def is_hashable(node: nodes.NodeNG) -> bool:
     try:
         for inferred in node.infer():
             if inferred is astroid.Uninferable:
+                return True
+            if not isinstance(inferred, WITH_IGETATTR):
                 return True
             hash_fn = next(inferred.igetattr("__hash__"))
             if hash_fn.parent is inferred:

--- a/tests/functional/u/unhashable_member.py
+++ b/tests/functional/u/unhashable_member.py
@@ -21,3 +21,4 @@ class Unhashable:
 {[1, 2, 3]}  # [unhashable-member]
 {"tomato": "tomahto"}
 {dict: {}}
+{lambda x: x: "tomato"}  # pylint: disable=unnecessary-lambda

